GVFS/GVFS/CommandLine/UpgradeVerb.cs

```diff
@@ -0,0 +1,309 @@
+            }
+
+            if (!this.Confirmed)
+            {
+                this.ReportInfoToConsole("Run `gvfs upgrade --confirm` to install it");
+                return true;
+            }
+
+            if (!this.TryRunInstaller(out error))
+            {
+                this.tracer.RelatedError($"{nameof(this.TryRunProductUpgrade)}: Could not launch installer. {error}");
+                this.Output.WriteLine(errorOutputFormat, "Could not launch installer. " + error);
+                return false;
+            }
+
+            return true;
+        }
+
+        private bool TryRunUpgradeChecks(
+            out Version latestVersion,
+            out bool installable,
+            out string error)
+        {
+            bool upgradeCheckSuccess = false;
+            bool upgradeInstallable = false;
+            string errorMessage = null;
+            Version version = null;
+
+            this.ShowStatusWhileRunning(
+                () =>
+                {
+                    upgradeCheckSuccess = this.TryCheckUpgradeAvailable(out version, out errorMessage);
+                    if (upgradeCheckSuccess && version != null)
+                    {
+                        upgradeInstallable = true;
+
+                        if (!this.TryCheckUpgradeInstallable(out errorMessage))
+                        {
+                            upgradeInstallable = false;
+                        }
+                    }
+
+                    return upgradeCheckSuccess;
+                },
+                 "Checking for GVFS upgrades",
+                suppressGvfsLogMessage: true);
+
+            latestVersion = version;
+            installable = upgradeInstallable;
+            error = errorMessage;
+
+            return upgradeCheckSuccess;
+        }
+
+        private bool TryRunInstaller(out string error)
+        {
+            string upgraderPath = null;
+            string errorMessage = null;
+
+            bool preUpgradeSuccess = this.ShowStatusWhileRunning(
+                () =>
+                {
+                    if (this.TryCopyUpgradeTool(out upgraderPath, out errorMessage) &&
+                        this.TryLaunchUpgradeTool(upgraderPath, out errorMessage))
+                    {
+                        return true;
+                    }
+
+                    return false;
+                },
+                "Launching upgrade tool",
+                suppressGvfsLogMessage: true);
+            
+            if (!preUpgradeSuccess)
+            {
+                error = errorMessage;
+                return false;
+            }
+
+            error = null;
+            return true;
+        }
+
+        private bool TryCopyUpgradeTool(out string upgraderExePath, out string error)
+        {
+            upgraderExePath = null;
+
+            this.tracer.RelatedInfo("Copying upgrade tool");
+
+            if (!this.upgrader.TryCreateToolsDirectory(out upgraderExePath, out error))

```

Comment:
Minor naming suggestion: `TryCreateToolsDirectory` sounds like it's just creating the tools directory, if the directory is also being copied\populated it would be helpful to include that in the name.

Reply:
renamed it to ```TrySetupToolsDirectory```